src/callbacks_step/ape_euler_coupling.jl

```diff
@@ -0,0 +1,111 @@
+
+end
+
+function Base.show(io::IO, ::MIME"text/plain", cb::DiscreteCallback{<:Any, <:ApeEulerCouplingCallback})
+
+end
+
+
+function ApeEulerCouplingCallback(cfl_ape::Real, cfl_euler::Real,
+                                  averaging_callback::DiscreteCallback{<:Any, <:AveragingCallback})
+
+  ape_euler_coupling = ApeEulerCouplingCallback{typeof(cfl_ape)}(StepsizeCallback(cfl_ape),
+                                                                 StepsizeCallback(cfl_euler),
+                                                                 averaging_callback, nothing)
+  condition = (u, t, integrator) -> true
+
+  return DiscreteCallback(condition, ape_euler_coupling, save_positions=(false, false),
+                          initialize=initialize!)
+end
+
+function initialize!(cb::DiscreteCallback{Condition,Affect!}, u_ode, t, integrator) where {Condition, Affect!<:ApeEulerCouplingCallback}
+  ape_euler_coupling = cb.affect!
+  semi = integrator.p
+  @unpack semi_ape, semi_euler = semi
+
+  # Set up ODE Integrator for Euler equations
+  tspan = integrator.sol.prob.tspan
+  alg = integrator.alg
+
+  ode_euler = semidiscretize(semi_euler, tspan)
+  ape_euler_coupling.integrator_euler = init(ode_euler, alg, save_everystep=false, dt=1.0) # dt will be overwritten
+
+  # Set mean values in u_ode according to `AveragingCallback`
+  u_ape = wrap_array(u_ode, semi_ape)
+  mesh, equations, solver, cache = mesh_equations_solver_cache(semi)
+  @unpack mean_values = ape_euler_coupling.averaging_callback.affect!
+  @views @. u_ape[4:5, .., :] = mean_values.v_mean
+  @views @. u_ape[6, .., :] = mean_values.c_mean
+  @views @. u_ape[7, .., :] = mean_values.rho_mean
+
+  # Calculate gradient of squared mean speed of sound for the q_cons source term
+  calc_gradient_c_mean_square!(semi.cache.grad_c_mean_sq, u_ape, mesh, equations, solver, cache)
+
+  # Adjust stepsize, advance the flow solver by one time step
+  cb.affect!(integrator)
+
+  return nothing
+end
+
+
+# This is called at the end of every time step and advances the Euler solution by one time step,
+# manages the time stepsize for both the APE and Euler solvers and calculates the acoustic sources
+# for the next APE time step
+function (ape_euler_coupling::ApeEulerCouplingCallback)(integrator_ape)
+  @unpack stepsize_callback_ape, stepsize_callback_euler, integrator_euler = ape_euler_coupling
+
+  @assert integrator_ape.t == integrator_euler.t
+
+  # Use the the minimum of the APE and Euler stepsizes for both solvers
+  stepsize_callback_ape(integrator_ape)
+  stepsize_callback_euler(integrator_euler)
+  dt = min(get_proposed_dt(integrator_ape), get_proposed_dt(integrator_euler))
+
+  set_proposed_dt!(integrator_ape, dt)
+  integrator_ape.opts.dtmax = dt
+  integrator_ape.dtcache = dt
+
+  set_proposed_dt!(integrator_euler, dt)
+  integrator_euler.opts.dtmax = dt
+  integrator_euler.dtcache = dt
+
+  # Advance the Euler solution by one step and check for errors
+  if !isfinished(integrator_euler)
```

Comment:
This whole process seems to be overly convoluted to me. It's a good prototype that works, so I wouldn't change it now, but it seems to me like we are abusing a lot of the OrdinaryDiffEq machinery here. We should somehow find another way to make this work in a cleaner fashion.

If not already done, can you maybe start an issue with open questions and known issues with this implementation, such that we can collect them in one place? First item from my side:
* Find a cleaner way to implement the coupled Euler-Acoustics evolution than manually using OrdinaryDiffEq internals